chore: Upgrade to syn v2.x and svgbob v0.7

jetuk opened the pull request:
Upgrade to syn v2.x and svgbob v0.7. Then address the resulting Clippy issues.

jetuk committed:
chore: Upgrade to syn v2.x and svgbob v0.7

Upgrade to syn v2.x and svgbob v0.7. Then address the resulting
Clippy issues.

jetuk committed:
chore(deps): Do not specify patch versions of dependencies.

yvt commented on rust-toolchain.toml:
```diff
@@ -1,2 +1,2 @@
 [toolchain]
-channel = "1.56"
+channel = "stable"
```
```suggestion
channel = "1.66"
```

yvt commented on Cargo.toml:
```diff
@@ -20,12 +20,12 @@ default = []
 proc-macro2 = "1"
-base64 = ">= 0.5.2, < 0.14"
-unicode-width = "0.1"
+base64 = "0.22"
```
```suggestion
base64 = ">= 0.21, < 0.23"
```

yvt commented on Cargo.toml:
```diff
@@ -20,12 +20,12 @@ default = []
 
 [dependencies]
-syn = "1.0.41"
+syn = "2.0"
```
```suggestion
syn = "2"
```

yvt commented on Cargo.toml:
```diff
@@ -20,12 +20,12 @@ default = []
-base64 = ">= 0.5.2, < 0.14"
-unicode-width = "0.1"
+base64 = "0.22"
+unicode-width = "0.2"
```
The difference between `^0.1` and `^0.2` is insignificant for us (afaik), but `svgbob ^0.7` still uses `^0.1`, so forcing the bump to `^0.2` would increase the build time.

```suggestion
unicode-width = ">= 0.1, < 0.3"
```

yvt reviewed (changes requested):
Sorry for the belated response! This looks good aside from a few minor issues.

Sympatron commented:
A release after this is merged would be nice. :)

jetuk committed:
chore(deps): Tidy-up dependencies, v1.66 toolchain.

Update dependency versions from review comments. Adjust
toolchain version to 1.66 and update in README.

jetuk commented:
Thanks for the review. I've made the updates in the latest commit and updated the Rust version in the README.

jetuk commented:
>  error: package `bumpalo v3.16.0` cannot be built because it requires rustc 1.73.0 or newer, while the currently active rustc version is 1.66.1

This is failing on the above. Would you prefer to bump the MSRV or downgrade `bumpalo`?

yvt commented:
Could you add `Cargo.lock` to the repo (removing it from `.gitignore`)?

Following the [default behavior](https://doc.rust-lang.org/cargo/faq.html#why-have-cargolock-in-version-control) of an older version of Cargo, `Cargo.lock` is currently excluded from the repository, which results in Cargo picking the latest dependencies on every CI run. Not only does this hurt build determinism, but it can also spontaneously break CI builds because Cargo can pick a crate version (`bumpalo v3.16.0` in this case) that is incompatible with the current compiler (the MSRV-aware resolver is still unstable). I think it's time we switched to the new behavior.

jetuk commented:
I've tried this and it requires going back to 3.14.0 which then has some build failures elsewhere in the dependency tree. I think it will be easier to bump MSRV to 1.73.0 to align with `bumpalo`. 

Also, I think the actions (derived from action-rs) are also unmaintained. I can update those to dtolnay actions which seem to be quite widely used now if you like?

yvt commented:
I see, let's bump to 1.73 then.

Upgrading the CI actions would be nice, but please do this in a separate PR.

jetuk committed:
chore(deps): Add Cargo.lock and bump MSRV to v1.73

Removes Cargo.lock from .gitignore.

jetuk committed:
Merge branch 'main' into syn-v2

jetuk committed:
chore(deps): Lock spade to v2.9.0

yvt commented on README.md:
```diff
@@ -5,7 +5,7 @@
 ASCII diagrams in doc comments as SVG images using [`svgbob`].
 
-*Requires Rust version 1.54 or later or equivalent nightly builds.*
+*Requires Rust version 1.66 or later or equivalent nightly builds.*
```
```suggestion
*Requires Rust version 1.73 or later or equivalent nightly builds.*
```

jetuk committed:
docs: Correct MSRV to 1.73 in README.

yvt commented:
Thanks!

Sympatron commented:
After this is merged now a release would be welcome :)

ROMemories commented:
Thanks for updating these dependencies! Could you cut a new release? `svgbobdoc` is one of the last users of `syn` v1 in our tree.

yvt commented:
@ROMemories Do you have any opinions on whether it should be a major version bump? On the one hand, bumping svgbob can break existing diagrams. But on the other hand, breaking diagrams could be insignificant compared to dependency hell.

Sympatron commented:
If svgbobdoc does not expose anything syn specific I don't think a major version bump is necessary. But if in doubt, you can use [cargo-semver-checks](https://github.com/obi1kenobi/cargo-semver-checks) to be sure.

ROMemories commented:
> Do you have any opinions on whether it should be a major version bump? On the one hand, bumping svgbob can break existing diagrams.

@yvt If it *can* break diagrams I do think there should be an incompatible version bump. However, looking at the [changelog of `svgbob`](https://github.com/ivanceras/svgbob/blob/master/Changelog.md#070), it seems to me that the only breaking change was about how the styles were handled, which I suppose should be entirely contained by `svgbobdoc`?

---

To be clear, we're not using `svgbobdoc` directly, it is only in our tree through `rlsf` ("we" meaning [ariel-os/ariel-os](https://github.com/ariel-os/ariel-os)).

Sympatron commented:
I totally missed that this PR updates `svgbob` as well... In that case it definitely has to be a major version bump.

What do you think about splitting this PR? Release 0.3.1 that just updates `syn` and then 0.4.0 with the update of `svgbob`.
That way crates like `rlsf` could profit immediately. Since rlsf wasn't updated in 3 years, I would rather not wait for them to do a release before downstream crates can get rid of `syn 1`…

I have a fork that just has the semver compatible parts of this PR [here](https://github.com/Sympatron/svgbobdoc/tree/syn-update), but I can't open a PR, because you can only open a PR against branches and not tags. You would have to create a branch at tag `0.3.0` for me to be able to open a PR.

yvt commented:
> If it _can_ break diagrams I do think there should be an incompatible version bump. However, looking at the [changelog of `svgbob`](https://github.com/ivanceras/svgbob/blob/master/Changelog.md#070), it seems to me that the only breaking change was about how the styles were handled, which I suppose should be entirely contained by `svgbobdoc`?

Technically, any changes to rendering (e.g., supporting three-quarter circles) can introduce breakage in existing diagrams because there are no input patterns reserved for future extension. `svgbob`'s semver scheme doesn't use this strict definition of breakage, as it can be seen that it has introduced rendering-changing changes in minor updates in the past (examples: [0.6.1](https://github.com/ivanceras/svgbob/commit/dfb72d701bb5bd0e59e8bee4daa86d6861d3ac46), [0.7.6](https://github.com/ivanceras/svgbob/commit/2f4f8f578a4fee2dbb398c3778229330829af88b)). This means if we followed this definition to the letter while keeping `svgbob` up-to-date, we'd have to pin the version of `svgbob` and do a major version bump even on a minor version bump of `svgbob`, which would negatively affect maintenance (lots of branches to maintain) and users' dependency graphs (multiple copies of `svgbob[doc]`).

For this reason, I'm somewhat inclined to tolerate some rendering-breaking changes as long as they don't break trivial diagrams, such as those made up of just hollow rectangles, perpendicular lines, and texts, meaning that the update to `svgbob` v0.7 is non-breaking.
